### PR TITLE
add Wmissing-prototypes to clang-tidy

### PR DESCRIPTION
**Pull request by cyyever**
This PR introduces **-Wmissing-prototypes** of clang-tidy to prevent further coding errors such as the one fixed by PR #96714. 

### <samp>🤖 Generated by Copilot at fd2cf2a</samp>

This pull request makes several internal functions static to improve performance and avoid name clashes. It also fixes some typos, formatting, and missing includes in various files. It adds a new .clang-tidy check to warn about missing prototypes for non-static functions.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -816,6 +816,8 @@ if(NOT MSVC)
   if("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
     string(APPEND CMAKE_CXX_FLAGS " -Wno-range-loop-analysis")
     string(APPEND CMAKE_CXX_FLAGS " -Wno-pass-failed")
+    string(APPEND CMAKE_CXX_FLAGS " -Wmissing-prototypes")
```
Comment (malfet) [inline CMakeLists.txt:819]: Please use 'append_cxx_flag_if_supported' rather conditioning it on clang

File: CMakeLists.txt [hunk 2]
```diff
@@ -816,6 +816,8 @@ if(NOT MSVC)
...
+    string(APPEND CMAKE_CXX_FLAGS " -Wno-error=missing-prototypes")
```
Comment (malfet) [inline CMakeLists.txt:820]: I assume this is temporary?

Comment (cyyever) [reply CMakeLists.txt]: We should use it permanently to check refactoring errors

**Comment by cyyever** [comment]
We can enable more optimizations by turning functions into static versions or fix some bugs by this warning option as indicated by commit #18b086bc1f98e5b45a7a0cd9434ea4ea7b0ded68

File: CMakeLists.txt [hunk 3]
```diff
@@ -814,6 +814,8 @@ if(NOT MSVC)
     string(APPEND CMAKE_CXX_FLAGS " -Wno-range-loop-analysis")
     string(APPEND CMAKE_CXX_FLAGS " -Wno-pass-failed")
   endif()
+  append_cxx_flag_if_supported("-Wmissing-prototypes" CMAKE_CXX_FLAGS)
+  append_cxx_flag_if_supported("-Wno-error=missing-prototypes" CMAKE_CXX_FLAGS)
```
Comment (malfet) [inline CMakeLists.txt:818]: Not sure what you are trying to achieve here. If We add `no-error` then one might not even be aware of the constraint and  free to ignore further regression.
```suggestion
  append_cxx_flag_if_supported("-Wmissing-prototypes" CMAKE_CXX_FLAGS)
```

Comment (cyyever) [reply CMakeLists.txt:818]: I am afraid that the current code base is not error free for Wmissing-prototypes after this PR even when we have applied some modifications, so I decide to temporarily treat it as a warning only option. After several further PRs to gradually check and fix the violating files then we can finally enable it

Comment (malfet) [reply CMakeLists.txt:818]: I propose to restrict it to some files  or only to `clang-tidy`, as there is nothing more annoying then 100 of lines of warnings, and it encourages anti-pattern of introducing new ones. Right now build logs on Linux and Mac are really compact which makes finding new errors easy. Let's keep them that way.

**Review by malfet — CHANGES_REQUESTED** [review]
I'm totally fine with adding a clang-tidy check, or adding it in cmake to say only `torch_cpu`, but I don't think it's a good idea to introduce an ignorable warning to the codebase.

**Comment by cyyever** [comment]
@malfet  OK, let's focus on clang-tidy only and revert CMake changes

File: torch/csrc/utils/byte_order.cpp [hunk 4]
```diff
@@ -462,7 +462,7 @@ void THP_encodeComplexFloatBuffer(
   }
 }
 
-void THP_encodeCompelxDoubleBuffer(
+void THP_encodeComplexDoubleBuffer(
```
Comment (albanD) [inline torch/csrc/utils/byte_order.cpp:465]: Is this never used? Should we just remove it?

Comment (cyyever) [reply torch/csrc/utils/byte_order.cpp:465]: Seems it is not used

File: torch/csrc/utils/verbose.cpp [hunk 5]
```diff
@@ -3,6 +3,7 @@
 
 namespace torch {
 
+// NOLINTNEXTLINE(clang-diagnostic-missing-prototypes)
```
[the remark this answers is not in the document]
Comment (cyyever) [reply torch/csrc/utils/verbose.cpp:6]: Because I can't find a header for this function

File: torch/csrc/utils/byte_order.cpp (hunk 4 above)
Comment (cyyever) [reply torch/csrc/utils/byte_order.cpp:466]: I have some fixes in 2b62567

**Review by albanD — APPROVED** [review]
Ho wow so many typos...
Thanks for the cleanup!

**Review by malfet — APPROVED** [review]
Looks good to me, but let's edit PR description a bit (and if you have bandwith, please feel free to submit a follow-up PR that enables `Wmissing-declarations` to select files in PyTorch using `set_source_files_properties`/`target_compile_options` and suppressing violations using `C10_DIAGNOSTIC_PUSH_AND_IGNORED_IF_DEFINED`

**Pull request closed without merging**